File: 454_4SumII/454_v1.cpp
class Solution {
public:
    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {
        unordered_map<int, int> hashTable;
        int count = 0;

        for (const int& a : A) {
            for (const int& b : B) {
                int value = a + b;
                auto iter = hashTable.find(value);
                if (iter == hashTable.end()) {
                    hashTable[value] = 0;
                }
                hashTable[value]++;
            }
        }


        for (const int& c : C) {
            for (const int& d : D) {
                int value = c + d;
                auto iter = hashTable.find(-value);
                if (iter != hashTable.end()) {
                    count += iter->second;
                }
            }
        }

        return count;
    }
};

File: 454_4SumII/454_v1.js
/**
 * @param {number[]} A
 * @param {number[]} B
 * @param {number[]} C
 * @param {number[]} D
 * @return {number}
 */
var fourSumCount = function(A, B, C, D) {
  const hashTable = {}
  let count = 0

  for (let i = 0; i < A.length; i++) {
      for (let j = 0; j < B.length; j++) {
          const value = A[i] + B[j]
          hashTable[value] = (hashTable[value] || 0) + 1
      }
  }

  for (let i = 0; i < C.length; i++) {
      for (let j = 0; j < D.length; j++) {
          const value = C[i] + D[j]
          count += (hashTable[-value] || 0)
      }
  }

  return count
};
